lib/AST/AlignTokens.cpp

```diff
@@ -2039,368 +1799,213 @@ Result<std::monostate, std::string> ASTImpl::AlignTokens(
-      tlds.begin(), tlds.end(),
-      [ast = ast.get()] (clang::Decl *a, clang::Decl *b) {
-        auto a_bounds = ast->DeclBounds(a);
-        auto b_bounds = ast->DeclBounds(b);
+static const std::string kEmptyParsed = "Parsed token range is empty";
+static const std::string kEmptyPrinted = "Printed token range is empty";
+static const std::string kMismatchedASTs = "Backing ASTs of parsed and printed token ranges don't match";
+static const std::string kMissingASTs = "No backing AST";
+}  // namespace
 
-        // If `a` starts first, put it first.
-        if (a_bounds.first < b_bounds.first) {
-          return true;
+// Create a new printed token range, where the token data is taken from `a`
+// but the token contexts are taken from `b`. This provides a mechanism of
+// relating parsed tokens back to AST nodes.
+Result<PrintedTokenRange, std::string>
+PrintedTokenRange::Align(const PrintedTokenRange &a,
+                         const PrintedTokenRange &b,
+                         bool maintain_provenance) {
+  if (!a) {
+    return kEmptyParsed;
+  }
 
-        } else if (a_bounds.first > b_bounds.first) {
-          return false;
+  if (!b) {
+    return kEmptyPrinted;
+  }
 
-        // If `b` encloses `a`, sort `b` first.
-        } else if (a_bounds.second < b_bounds.second) {
-          return false;
+  if (a.impl->ast.get() != b.impl->ast.get()) {
+    return kMismatchedASTs;
+  }
 
-        // If `a` encloses `b`, then sort `a` first.
-        } else if (a_bounds.second > b_bounds.second) {
-          return true;
+  if (!a.impl->ast) {
+    return kMissingASTs;
+  }
 
-        // Keep the relative order from `tlds`.
-        } else {
-          return false;
-        }
-      });
+  if (a.impl == b.impl && maintain_provenance) {
+    return a;
+  }
 
-  std::unordered_multimap<const void *, TokenContextIndex> data_to_context;
-  std::vector<TokenContextIndex> context_map;
-  std::vector<clang::Decl *> tld_group;
-  std::vector<const clang::Decl *> parentage;
-  std::vector<TokenPrinterContext> context_stack;
-  std::string data;
-  auto &ast_context = ast->tu->getASTContext();
-
-  for (auto tld_it = tlds.begin(), tld_end = tlds.end(); tld_it != tld_end; ) {
-    clang::Decl *decl = *tld_it;
-    clang::Decl *&containing_decl = ast->lexically_containing_decl[decl];
-    if (!containing_decl) {
-      containing_decl = decl;
-    }
-
-    tld_group.clear();
-    tld_group.push_back(containing_decl);
-    for (; tld_it != tld_end; ++tld_it) {
-      clang::Decl *next_decl = *tld_it;
-      if (containing_decl == ast->lexically_containing_decl[next_decl]) {
-        if (next_decl != containing_decl) {
-          tld_group.push_back(next_decl);
-        }
-      } else {
-        break;
-      }
-    }
+  auto new_impl = std::make_shared<PrintedTokenRangeImpl>(
+      a.impl->ast->ci->getASTContext());
+  new_impl->ast = a.impl->ast;
+  new_impl->tokens.reserve(a.impl->tokens.size());
+  new_impl->data = a.impl->data;
```

Comment:
Ctor.

Reply:
I ended up making this API no return an aligned range, but mutate the inputs in place, which actually was partially always happening. Now the prototype/docs better communicate that.